## Company Bankruptcy Prediction

In [1]:
import numpy as np 
import pandas as pd 

In [43]:
df=pd.read_csv("../data/train.csv", low_memory=False)

In [3]:
train = df.apply(pd.to_numeric, errors='coerce')
train.head()

id  forecasting period  net_profit_over_total_assets  \
0   0                   4                      0.141939   
1   1                   4                      0.147560   
2   2                   1                     -0.024542   
3   3                   2                      0.159317   
4   4                   4                      0.004404   

   total_liabilities_over_total_assets  working_capital_over_total_assets  \
0                             0.450292                           0.021422   
1                             0.234840                           0.708335   
2                             0.311544                           0.067831   
3                             0.777017                           0.207431   
4                             0.577419                          -0.104435   

   current_assets_over_short_term_liabilities  \
0                                    1.052987   
1                                    4.016018   
2                                    1.255247   
3                                    1.267310   
4                                    0.786225   

   cash_plus_short_term_securities_plus_receivables_minus_short_term_liabilities_over_operating_expenses_minus_depreciation_times_365  \
0                                         -28.529084                                                                                    
1                                         119.445611                                                                                    
2                                         -15.019923                                                                                    
3                                          -4.892918                                                                                    
4                                        -116.911618                                                                                    

   retained_earnings_over_total_assets  EBIT_over_total_assets  \
0                             0.141934                0.172629   
1                             0.457090                0.186070   
2                            -0.657003               -0.024537   
3                            -0.000005                0.159321   
4                             0.056561                0.004249   

   book_value_of_equity_over_total_liabilities  ...  \
0                                     0.990337  ...   
1                                     3.206445  ...   
2                                     2.209000  ...   
3                                     0.288177  ...   
4                                     0.692166  ...   

   sales_minus_cost_of_products_sold_over_sales  \
0                                      0.079572   
1                                      0.095218   
2                                      0.109035   
3                                      0.102971   
4                                      0.016446   

   current_assets_minus_inventory_minus_short_term_liabilities_over_sales_minus_gross_profit_minus_depreciation  \
0                                           0.318286                                                              
1                                           0.195943                                                              
2                                          -0.035634                                                              
3                                           0.714528                                                              
4                                           0.011019                                                              

   total_costs_overtotal_sales  long_term_liabilities_over_equity  \
0                     0.927687                           0.106818   
1                     0.907056                          -0.000029   
2                     0.929825                           0.055331   
3                     0.937895                          -0.000077

In [4]:
train.columns

Index(['id', 'forecasting period', 'net_profit_over_total_assets',
       'total_liabilities_over_total_assets',
       'working_capital_over_total_assets',
       'current_assets_over_short_term_liabilities',
       'cash_plus_short_term_securities_plus_receivables_minus_short_term_liabilities_over_operating_expenses_minus_depreciation_times_365',
       'retained_earnings_over_total_assets', 'EBIT_over_total_assets',
       'book_value_of_equity_over_total_liabilities',
       'sales_over_total_assets', 'equity_over_total_assets',
       'gross_profit_plus_extraordinary_items_plus_financial_expenses_over_total_assets',
       'gross_profit_over_short_term_liabilities',
       'gross_profit_plus_depreciation_over_sales',
       'gross_profit_plus_interest_over_total_assets',
       'total_liabilities_times_365_over_gross_profit_plus_depreciation',
       'gross_profit_plus_depreciation_over_total_liabilities',
       'total_assets_over_total_liabilities', 'gross_profit_over_total_asse

### Exploratory Analysis for Variables

In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25121 entries, 0 to 25120
Data columns (total 67 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   id                                                                                                                                  25121 non-null  int64  
 1   forecasting period                                                                                                                  25121 non-null  int64  
 2   net_profit_over_total_assets                                                                                                        25118 non-null  float64
 3   total_liabilities_over_total_assets                                              

#### Dependent Variable

In [6]:
sum(train["class"].isna())

0

In [7]:
sum(train["class"].isnull())

0

So clearly there is no missing values in the dependent variable.

In [8]:
train["class"].unique() ## checking for unique values

array([1, 0], dtype=int64)

In [9]:
train.groupby(['class'])[['class']].count()

class
class       
0      23995
1       1126

In [10]:
round(len(train["class"][train["class"] == 1])/len(train["class"]), 4) ## Percentage of Bankruptcy Cases

0.0448

In [11]:
round(len(train["class"][train["class"] == 1])/len(train["class"][train["class"] == 0]), 4) ## Odds Ratio

0.0469

The total number of Bankruptcy cases is 1126 and it comprises 0.0448 percentage of the whole data.

### Missing value imputation

In [12]:
train.isna().sum().sort_values(ascending=False)

current_assets_minus_inventories_over_long_term_liabilities    11153
sales_n_over_sales_n_minus_1                                    3911
profit_on_operating_activities_over_financial_expenses          1600
sales_over_inventory                                            1282
net_profit_over_inventory                                       1279
                                                               ...  
equity_over_total_assets                                           3
working_capital                                                    1
id                                                                 0
forecasting period                                                 0
class                                                              0
Length: 67, dtype: int64

In [13]:
for i in train:
    train.loc[train.loc[:,i].isnull(),i]=train.loc[:,i].mean()

In [14]:
train.isna().sum().sort_values(ascending=False)

id                                                                     0
EBITDA_profit_on_operating_activities_minus_depreciation_over_sales    0
profit_on_sales_over_total_assets                                      0
total_sales_over_total_assets                                          0
current_assets_minus_inventories_over_long_term_liabilities            0
                                                                      ..
profit_on_operating_activities_over_financial_expenses                 0
working_capital_over_fixed_assets                                      0
logarithm_of_total_assets                                              0
total_liabilities_minus_cash_over_sales                                0
class                                                                  0
Length: 67, dtype: int64

We can see that almost all of the columns has missing values. So before we proceed with the analysis we have to impute these values.

In [15]:
train.describe()

id  forecasting period  net_profit_over_total_assets  \
count  25121.000000        25121.000000                  25121.000000   
mean   12560.000000            2.608574                      0.046883   
std     7251.952392            1.063099                      1.821587   
min        0.000000            1.000000                   -256.889998   
25%     6280.000000            2.000000                      0.003354   
50%    12560.000000            3.000000                      0.049838   
75%    18840.000000            4.000000                      0.131410   
max    25120.000000            4.000000                     52.651999   

       total_liabilities_over_total_assets  working_capital_over_total_assets  \
count                         25121.000000                       25121.000000   
mean                              0.611913                           0.098553   
std                               5.933092                           5.907783   
min                             -72.162006                        -479.959994   
25%                               0.271085                           0.016667   
50%                               0.476985                           0.191632   
75%                               0.693518                           0.401742   
max                             480.960001                          22.769001   

       current_assets_over_short_term_liabilities  \
count                                25121.000000   
mean                                     6.285753   
std                                    339.445391   
min                                     -0.044851   
25%                                      1.039445   
50%                                      1.557344   
75%                                      2.784023   
max                                  53433.000440   

       cash_plus_short_term_securities_plus_receivables_minus_short_term_liabilities_over_operating_expenses_minus_depreciation_times_365  \
count                                       2.512100e+04                                                                                    
mean                                       -9.125936e+01                                                                                    
std                                         1.826400e+04                                                                                    
min                                        -1.670200e+06                                                                                    
25%                                        -5.106407e+01                                                                                    
50%                                        -1.874916e+00                                                                                    
75%                                         5.105802e+01                                                                                    
max                                         1.034100e+06                                                                                    

       retained_earnings_over_total_assets  EBIT_over_total_assets  \
count                         25121.000000            25121.000000   
mean                             -0.067728                0.147535   
std                               6.872710                5.834939   
min                            -508.409992             -189.559992   
25%                              -0.000010                0.005652   
50%                               0.000003                0.060084   
75%                               0.085011                0.153183   
max                             322.199997              649.230003   

       book_value_of_equity_over_total_liabilities  ...  \
count                                 25121.000000  ...   
mean                                     15.153162  ...   
std                                     610.876764  ...   
min           

### Train and validation split

In [16]:
X = train.iloc[:, 1:-1]
y = train.iloc[:,  -1]

In [17]:
X.head()

forecasting period  net_profit_over_total_assets  \
0                 4.0                      0.141939   
1                 4.0                      0.147560   
2                 1.0                     -0.024542   
3                 2.0                      0.159317   
4                 4.0                      0.004404   

   total_liabilities_over_total_assets  working_capital_over_total_assets  \
0                             0.450292                           0.021422   
1                             0.234840                           0.708335   
2                             0.311544                           0.067831   
3                             0.777017                           0.207431   
4                             0.577419                          -0.104435   

   current_assets_over_short_term_liabilities  \
0                                    1.052987   
1                                    4.016018   
2                                    1.255247   
3                                    1.267310   
4                                    0.786225   

   cash_plus_short_term_securities_plus_receivables_minus_short_term_liabilities_over_operating_expenses_minus_depreciation_times_365  \
0                                         -28.529084                                                                                    
1                                         119.445611                                                                                    
2                                         -15.019923                                                                                    
3                                          -4.892918                                                                                    
4                                        -116.911618                                                                                    

   retained_earnings_over_total_assets  EBIT_over_total_assets  \
0                             0.141934                0.172629   
1                             0.457090                0.186070   
2                            -0.657003               -0.024537   
3                            -0.000005                0.159321   
4                             0.056561                0.004249   

   book_value_of_equity_over_total_liabilities  sales_over_total_assets  ...  \
0                                     0.990337                 1.087356  ...   
1                                     3.206445                 1.093404  ...   
2                                     2.209000                 0.825578  ...   
3                                     0.288177                 2.283532  ...   
4                                     0.692166                 1.013605  ...   

   working_capital  sales_minus_cost_of_products_sold_over_sales  \
0       800.542408                                      0.079572   
1     24093.000493                                      0.095218   
2       851.179200                                      0.109035   
3      2577.614840                                      0.102971   
4    -11887.079853                                      0.016446   

   current_assets_minus_inventory_minus_short_term_liabilities_over_sales_minus_gross_profit_minus_depreciation  \
0                                           0.318286                                                              
1                                           0.195943                                                              
2                                          -0.035634                                                              
3                                           0.714528                                                              
4                                           0.011019                                                              

   total_costs_overtotal_sales  long_term_liabilities_over_equity  \
0                     0.927687              

In [18]:
y.head()

0    1.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: class, dtype: float64

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=341)

### Test data preprocessing

In [44]:
df1 = pd.read_csv("../data/test.csv", low_memory=False)
test = df1.apply(pd.to_numeric, errors='coerce')
test.isna().sum().sort_values(ascending=False)

current_assets_minus_inventories_over_long_term_liabilities    5283
sales_n_over_sales_n_minus_1                                   1840
profit_on_operating_activities_over_financial_expenses          773
sales_over_inventory                                            602
net_profit_over_inventory                                       600
                                                               ... 
sales_over_total_assets                                           1
long_term_liabilities_over_equity                                 1
forecasting period                                                0
working_capital                                                   0
id                                                                0
Length: 66, dtype: int64

In [21]:
for i in test:
    test.loc[test.loc[:,i].isnull(),i]=test.loc[:,i].mean()

In [22]:
test.isna().sum().sort_values(ascending=False)

id                                                                            0
EBITDA_profit_on_operating_activities_minus_depreciation_over_total_assets    0
operating_expenses_over_total_liabilities                                     0
profit_on_sales_over_total_assets                                             0
total_sales_over_total_assets                                                 0
                                                                             ..
net_profit_plus_depreciation_over_total_liabilities                           0
profit_on_operating_activities_over_financial_expenses                        0
working_capital_over_fixed_assets                                             0
logarithm_of_total_assets                                                     0
sales_over_fixed_assets                                                       0
Length: 66, dtype: int64

In [23]:
test1 = test.iloc[:, 1:]
test1.head()

forecasting period  net_profit_over_total_assets  \
0                 4.0                      0.087054   
1                 2.0                     -0.210390   
2                 1.0                      0.755349   
3                 2.0                      0.090151   
4                 4.0                      0.023989   

   total_liabilities_over_total_assets  working_capital_over_total_assets  \
0                             0.504697                          -0.213442   
1                             0.631828                           0.228799   
2                             0.208157                           0.525567   
3                             0.539555                           0.220522   
4                             0.002572                           0.987142   

   current_assets_over_short_term_liabilities  \
0                                    0.527201   
1                                    1.369950   
2                                    3.525034   
3                                    1.409651   
4                                  385.569944   

   cash_plus_short_term_securities_plus_receivables_minus_short_term_liabilities_over_operating_expenses_minus_depreciation_times_365  \
0                                        -945.320442                                                                                    
1                                         -63.351273                                                                                    
2                                          95.936870                                                                                    
3                                          -8.775940                                                                                    
4                                        1704.227987                                                                                    

   retained_earnings_over_total_assets  EBIT_over_total_assets  \
0                         2.456644e-07                0.091917   
1                        -3.844551e-02               -0.258109   
2                         7.305173e-03                0.755342   
3                        -3.165242e-01                0.090153   
4                         1.438131e-05                0.029510   

   book_value_of_equity_over_total_liabilities  sales_over_total_assets  ...  \
0                                     0.982293                 1.642851  ...   
1                                     0.582886                 1.670711  ...   
2                                     3.804564                 2.604779  ...   
3                                     0.852674                 2.567658  ...   
4                                   388.589564                 0.210381  ...   

   working_capital  sales_minus_cost_of_products_sold_over_sales  \
0    -19987.124701                                      0.322386   
1        18.621606                                     -0.126736   
2      1440.011406                                      0.337905   
3       503.854614                                      0.220679   
4     12841.024909                                      0.051816   

   current_assets_minus_inventory_minus_short_term_liabilities_over_sales_minus_gross_profit_minus_depreciation  \
0                                           0.175775                                                              
1                                          -0.571445                                                              
2                                           0.953913                                                              
3                                           0.195811                                                              
4                                           0.024058                                                              

   total_costs_overtotal_sales  long_term_liabilities_over_equity  \
0                     0.706656              

Now that we have imputed both train and test using KNN we can start the model building. 

### Model Building

#### Logistic Regression

In [24]:
from sklearn import preprocessing
scaler = preprocessing.StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [25]:
from sklearn.feature_selection import SelectKBest
k = 12
prep = SelectKBest(k=k)
X_train = prep.fit_transform(X_train, y_train)
X_test = prep.fit_transform(X_test, y_test)

In [28]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import recall_score
from sklearn.metrics import accuracy_score, f1_score
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(n_jobs = -1, random_state=0, max_iter = 100000, solver = 'saga')
lr.fit(X_train, y_train)
y_train_pred = lr.predict(X_train)
y_test_pred = lr.predict(X_test)
print(f1_score(y_train, y_train_pred))
print(f1_score(y_test, y_test_pred))

0.004944375772558715
0.01238390092879257


In [ ]:
from sklearn.model_selection import GridSearchCV
cv = 3
hyperparameters = {'penalty': [ 'l2','none'], 'C': [0.1, 1, 10, 100]}
regression_grid = GridSearchCV(lr, hyperparameters, scoring='accuracy', cv=cv, verbose=0, n_jobs=-1)
regression_grid.fit(X_train, y_train)

In [68]:
print('Regression Classifier Best Parameters')
print(regression_grid.best_params_)

Regression Classifier Best Parameters
{'C': 0.1, 'penalty': 'l2'}


##### Model evaluation

In [69]:
y_test_pred = regression_grid.best_estimator_.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7217    0]
 [ 319    1]]
              precision    recall  f1-score   support

           0       0.96      1.00      0.98      7217
           1       1.00      0.00      0.01       320

    accuracy                           0.96      7537
   macro avg       0.98      0.50      0.49      7537
weighted avg       0.96      0.96      0.94      7537



### SVM

In [70]:
from sklearn import preprocessing
scaler = preprocessing.StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [71]:
from sklearn.feature_selection import SelectKBest
k = 10
prep = SelectKBest(k=k)
X_train = prep.fit_transform(X_train, y_train)
X_test = prep.fit_transform(X_test, y_test)

In [ ]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import recall_score
from sklearn.metrics import accuracy_score
from sklearn.svm import LinearSVC

clf = LinearSVC(random_state=0, C = 0.01, max_iter = 100000)
clf.fit(X_train, y_train)
y_train_pred = clf.predict(X_train)
y_test_pred = clf.predict(X_test)
print(f1_score(y_train, y_train_pred))
print(f1_score(y_test, y_test_pred)) 

In [73]:
from sklearn.model_selection import GridSearchCV
cv = 5
hyperparameters = {'C': [0.01, 0.1, 1, 10, 100]}
regression_grid = GridSearchCV(LinearSVC(random_state=92, max_iter=10000), hyperparameters, scoring='accuracy', cv=cv, verbose=0, n_jobs=-1)
regression_grid.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=LinearSVC(max_iter=10000, random_state=92),
             n_jobs=-1, param_grid={'C': [0.01, 0.1, 1, 10, 100]},
             scoring='accuracy')

In [76]:
print('Regression Classifier Best Parameters')
print(regression_grid.best_params_)

Regression Classifier Best Parameters
{'C': 0.01}


In [77]:
y_test_pred = regression_grid.best_estimator_.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7216    1]
 [ 318    2]]
              precision    recall  f1-score   support

           0       0.96      1.00      0.98      7217
           1       0.67      0.01      0.01       320

    accuracy                           0.96      7537
   macro avg       0.81      0.50      0.50      7537
weighted avg       0.95      0.96      0.94      7537



### Decision tree

In [24]:
X = train.iloc[:, 1:-1]

from sklearn.model_selection import train_test_split
y = train[['class']].values.ravel()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [26]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import recall_score
from sklearn.metrics import accuracy_score, f1_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import plot_tree

dt = DecisionTreeClassifier(random_state=92, max_depth=2)
dt.fit(X_train, y_train)
y_train_pred = dt.predict(X_train)
y_test_pred = dt.predict(X_test)
print(f1_score(y_train, y_train_pred))
print(f1_score(y_test, y_test_pred))

0.403770620581304
0.4110429447852761


In [146]:
dt = DecisionTreeClassifier(random_state=0, max_depth=6)
dt.fit(X_train, y_train)
y_train_pred = dt.predict(X_train)
y_test_pred = dt.predict(X_test)
print(f1_score(y_train, y_train_pred))
print(f1_score(y_test, y_test_pred))

0.6046114432109307
0.4890829694323144


In [73]:
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7191   26]
 [ 208  112]]
              precision    recall  f1-score   support

         0.0       0.97      1.00      0.98      7217
         1.0       0.81      0.35      0.49       320

    accuracy                           0.97      7537
   macro avg       0.89      0.67      0.74      7537
weighted avg       0.97      0.97      0.96      7537



### ANN

In [18]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import recall_score
from sklearn.metrics import accuracy_score, f1_score
from sklearn import preprocessing

In [19]:
scaler = preprocessing.StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [20]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(hidden_layer_sizes=(100, 50, 25, 2), random_state=0, beta_1 = 0.99) # alpha is for regularization (L2 regularization)
mlp.fit(X_train, y_train)
y_train_pred = mlp.predict(X_train)
y_test_pred = mlp.predict(X_test)
print('Training accuracy:'.format(), f1_score(y_train, y_train_pred))
print('Test accuracy:'.format(), f1_score(y_test, y_test_pred))

Training accuracy: 0.8376830044557607
Test accuracy: 0.43174603174603177


In [85]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(hidden_layer_sizes=(100, 50, 25, 2), random_state=0, beta_1 = 0.99) # alpha is for regularization (L2 regularization)
mlp.fit(X_train, y_train)
y_train_pred = mlp.predict(X_train)
y_test_pred = mlp.predict(X_test)
print('Training accuracy:'.format(), f1_score(y_train, y_train_pred))
print('Test accuracy:'.format(), f1_score(y_test, y_test_pred))

Training accuracy: 0.8376830044557607
Test accuracy: 0.43174603174603177


In [87]:
scaler = preprocessing.StandardScaler().fit(test1)
test2 = scaler.transform(test1)
grid_predictions = mlp.predict(test2)
pred=pd.DataFrame(grid_predictions, index = test.id)
pred.value_counts()

0.0    11306
1.0     1068
dtype: int64

In [88]:
import os
os.chdir(r'../data')

pred.to_csv(r"submission_ann.csv")

In [18]:
from sklearn.model_selection import GridSearchCV
cv = 3
hyperparameters = {'alpha': [0.0001, 0.001, 0.01, 0.1]}
mlp_grid = GridSearchCV(mlp, hyperparameters, scoring='accuracy', cv=cv, verbose=0, n_jobs=-1)
mlp_grid.fit(X_train, y_train)

GridSearchCV(cv=3,
             estimator=MLPClassifier(alpha=1e-05, hidden_layer_sizes=(5, 2),
                                     max_iter=5000, random_state=0),
             n_jobs=-1, param_grid={'alpha': [0.0001, 0.001, 0.01, 0.1]},
             scoring='accuracy')

In [19]:
print('Regression Classifier Best Parameters:')
print(mlp_grid.best_params_)

Regression Classifier Best Parameters:
{'alpha': 0.01}


In [23]:
y_test_pred = mlp_grid.best_estimator_.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7169   48]
 [ 240   80]]
              precision    recall  f1-score   support

         0.0       0.97      0.99      0.98      7217
         1.0       0.62      0.25      0.36       320

    accuracy                           0.96      7537
   macro avg       0.80      0.62      0.67      7537
weighted avg       0.95      0.96      0.95      7537



In [188]:
### Ridge classification

from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import RidgeClassifier
clf = RidgeClassifier().fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[4792    2]
 [ 230    1]]
              precision    recall  f1-score   support

         0.0       0.95      1.00      0.98      4794
         1.0       0.33      0.00      0.01       231

    accuracy                           0.95      5025
   macro avg       0.64      0.50      0.49      5025
weighted avg       0.93      0.95      0.93      5025



In [184]:
#### SGD

import numpy as np
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

clf = make_pipeline(StandardScaler(),SGDClassifier(max_iter=1000, tol=1e-3,class_weight='balanced'))
clf.fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[3621 1173]
 [ 109  122]]
              precision    recall  f1-score   support

         0.0       0.97      0.76      0.85      4794
         1.0       0.09      0.53      0.16       231

    accuracy                           0.74      5025
   macro avg       0.53      0.64      0.50      5025
weighted avg       0.93      0.74      0.82      5025



In [186]:
from sklearn.linear_model import PassiveAggressiveClassifier

clf = PassiveAggressiveClassifier(max_iter=1000, random_state=0,tol=1e-3)
clf.fit(X_train, y_train)
PassiveAggressiveClassifier(random_state=0)

y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[4547  247]
 [ 205   26]]
              precision    recall  f1-score   support

         0.0       0.96      0.95      0.95      4794
         1.0       0.10      0.11      0.10       231

    accuracy                           0.91      5025
   macro avg       0.53      0.53      0.53      5025
weighted avg       0.92      0.91      0.91      5025



In [181]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

clf = LinearDiscriminantAnalysis()
clf.fit(X_train, y_train)
LinearDiscriminantAnalysis()
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[4789    5]
 [ 227    4]]
              precision    recall  f1-score   support

         0.0       0.95      1.00      0.98      4794
         1.0       0.44      0.02      0.03       231

    accuracy                           0.95      5025
   macro avg       0.70      0.51      0.50      5025
weighted avg       0.93      0.95      0.93      5025



In [179]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
clf = QuadraticDiscriminantAnalysis()
clf.fit(X_train, y_train)
QuadraticDiscriminantAnalysis()
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[ 338 4456]
 [  13  218]]
              precision    recall  f1-score   support

         0.0       0.96      0.07      0.13      4794
         1.0       0.05      0.94      0.09       231

    accuracy                           0.11      5025
   macro avg       0.50      0.51      0.11      5025
weighted avg       0.92      0.11      0.13      5025



In [177]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import NuSVC
clf = make_pipeline(StandardScaler(), NuSVC(nu=0.01,class_weight='balanced'))
clf.fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[1621 3173]
 [  48  183]]
              precision    recall  f1-score   support

         0.0       0.97      0.34      0.50      4794
         1.0       0.05      0.79      0.10       231

    accuracy                           0.36      5025
   macro avg       0.51      0.57      0.30      5025
weighted avg       0.93      0.36      0.48      5025



In [ ]:
from sklearn.svm import LinearSVC
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_classification
clf = make_pipeline(StandardScaler(),LinearSVC(random_state=0, tol=1e-5,max_iter=1000000,class_weight='balanced'))
clf.fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

In [27]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [ ]:
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
gpc = GaussianProcessClassifier(random_state=0).fit(X_train, y_train)
y_test_pred = gpc.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

In [12]:
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7201   16]
 [ 318    2]]
              precision    recall  f1-score   support

         0.0       0.96      1.00      0.98      7217
         1.0       0.11      0.01      0.01       320

    accuracy                           0.96      7537
   macro avg       0.53      0.50      0.49      7537
weighted avg       0.92      0.96      0.94      7537



In [299]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
clf = GaussianNB()
clf.fit(X_train, y_train)
GaussianNB()
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[ 250 6967]
 [  14  306]]
              precision    recall  f1-score   support

         0.0       0.95      0.03      0.07      7217
         1.0       0.04      0.96      0.08       320

    accuracy                           0.07      7537
   macro avg       0.49      0.50      0.07      7537
weighted avg       0.91      0.07      0.07      7537



In [17]:
from sklearn.naive_bayes import BernoulliNB
clf = BernoulliNB()
clf.fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[5764 1453]
 [ 176  144]]
              precision    recall  f1-score   support

         0.0       0.97      0.80      0.88      7217
         1.0       0.09      0.45      0.15       320

    accuracy                           0.78      7537
   macro avg       0.53      0.62      0.51      7537
weighted avg       0.93      0.78      0.85      7537



In [102]:
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
clf = DecisionTreeClassifier(random_state=0)
clf.fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7023  194]
 [ 155  165]]
              precision    recall  f1-score   support

         0.0       0.98      0.97      0.98      7217
         1.0       0.46      0.52      0.49       320

    accuracy                           0.95      7537
   macro avg       0.72      0.74      0.73      7537
weighted avg       0.96      0.95      0.95      7537



In [103]:
grid_predictions = clf.predict(test1)
pred=pd.DataFrame(grid_predictions, index = test.id)
pred.value_counts()

0.0    11757
1.0      617
dtype: int64

In [77]:
import os
os.chdir(r'../data')

pred.to_csv(r"submission_dt.csv")

In [23]:
from sklearn.svm import SVC
from sklearn.ensemble import BaggingClassifier
from sklearn.datasets import make_classification
clf = BaggingClassifier(base_estimator=SVC(class_weight='balanced'),n_estimators=10, random_state=0).fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[4715 2502]
 [ 141  179]]
              precision    recall  f1-score   support

         0.0       0.97      0.65      0.78      7217
         1.0       0.07      0.56      0.12       320

    accuracy                           0.65      7537
   macro avg       0.52      0.61      0.45      7537
weighted avg       0.93      0.65      0.75      7537



In [247]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
clf =  RandomForestClassifier(n_estimators=25, random_state=0, bootstrap = False)
clf.fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7198   19]
 [ 209  111]]
              precision    recall  f1-score   support

         0.0       0.97      1.00      0.98      7217
         1.0       0.85      0.35      0.49       320

    accuracy                           0.97      7537
   macro avg       0.91      0.67      0.74      7537
weighted avg       0.97      0.97      0.96      7537



In [248]:
grid_predictions = clf.predict(test1)
pred=pd.DataFrame(grid_predictions, index = test.id)
pred.value_counts()

0.0    12139
1.0      235
dtype: int64

In [249]:
import os
os.chdir(r'../data')

pred.to_csv(r"submission_bt.csv")

In [28]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.datasets import make_classification
clf = ExtraTreesClassifier(n_estimators=100, random_state=0)
clf.fit(X_train, y_train)
ExtraTreesClassifier(random_state=0)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7199   18]
 [ 306   14]]
              precision    recall  f1-score   support

         0.0       0.96      1.00      0.98      7217
         1.0       0.44      0.04      0.08       320

    accuracy                           0.96      7537
   macro avg       0.70      0.52      0.53      7537
weighted avg       0.94      0.96      0.94      7537



In [85]:
from sklearn.ensemble import GradientBoostingClassifier

clf = GradientBoostingClassifier(random_state=0, learning_rate = 0.2, n_estimators = 3500).fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7196   21]
 [ 148  172]]
              precision    recall  f1-score   support

         0.0       0.98      1.00      0.99      7217
         1.0       0.89      0.54      0.67       320

    accuracy                           0.98      7537
   macro avg       0.94      0.77      0.83      7537
weighted avg       0.98      0.98      0.97      7537



In [90]:
grid_predictions = clf.predict(test1)
pred=pd.DataFrame(grid_predictions, index = test.id)
pred.value_counts()

0.0    12075
1.0      299
dtype: int64

In [91]:
import os
os.chdir(r'../data')

pred.to_csv(r"submission_gb2.csv")

In [86]:
from sklearn.ensemble import GradientBoostingClassifier

clf1 = GradientBoostingClassifier(random_state=0, learning_rate = 0.2, n_estimators = 3500, min_samples_leaf = 9).fit(X_train, y_train)
y_test_pred = clf1.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7203   14]
 [ 149  171]]
              precision    recall  f1-score   support

         0.0       0.98      1.00      0.99      7217
         1.0       0.92      0.53      0.68       320

    accuracy                           0.98      7537
   macro avg       0.95      0.77      0.83      7537
weighted avg       0.98      0.98      0.98      7537



In [108]:
from sklearn.ensemble import GradientBoostingClassifier

clf1 = GradientBoostingClassifier(random_state=0, n_estimators = 3500, min_samples_leaf = 8).fit(X_train, y_train)
y_test_pred = clf1.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7203   14]
 [ 152  168]]
              precision    recall  f1-score   support

         0.0       0.98      1.00      0.99      7217
         1.0       0.92      0.53      0.67       320

    accuracy                           0.98      7537
   macro avg       0.95      0.76      0.83      7537
weighted avg       0.98      0.98      0.98      7537



In [87]:
grid_predictions = clf1.predict(test1)
pred=pd.DataFrame(grid_predictions, index = test.id)
pred.value_counts()

0.0    12085
1.0      289
dtype: int64

In [88]:
import os
os.chdir(r'../data')

pred.to_csv(r"submission_gb1.csv")

In [109]:
from sklearn.ensemble import GradientBoostingClassifier

clf = GradientBoostingClassifier(random_state=0, n_estimators = 600, min_samples_leaf = 8).fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7213    4]
 [ 155  165]]
              precision    recall  f1-score   support

         0.0       0.98      1.00      0.99      7217
         1.0       0.98      0.52      0.67       320

    accuracy                           0.98      7537
   macro avg       0.98      0.76      0.83      7537
weighted avg       0.98      0.98      0.98      7537



In [110]:
grid_predictions = clf.predict(test1)
pred=pd.DataFrame(grid_predictions, index = test.id)
pred.value_counts()

0.0    12122
1.0      252
dtype: int64

In [111]:
import os
os.chdir(r'../data')

pred.to_csv(r"submission_gb3.csv")

In [107]:
from sklearn.ensemble import GradientBoostingClassifier

clf = GradientBoostingClassifier(random_state=0, n_estimators = 600, min_samples_leaf = 8, max_depth = 5).fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7209    8]
 [ 157  163]]
              precision    recall  f1-score   support

         0.0       0.98      1.00      0.99      7217
         1.0       0.95      0.51      0.66       320

    accuracy                           0.98      7537
   macro avg       0.97      0.75      0.83      7537
weighted avg       0.98      0.98      0.97      7537



In [90]:
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.datasets import load_iris
clf = HistGradientBoostingClassifier().fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7213    4]
 [ 165  155]]
              precision    recall  f1-score   support

         0.0       0.98      1.00      0.99      7217
         1.0       0.97      0.48      0.65       320

    accuracy                           0.98      7537
   macro avg       0.98      0.74      0.82      7537
weighted avg       0.98      0.98      0.97      7537



In [28]:
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.datasets import load_iris
clf = HistGradientBoostingClassifier(random_state=252, l2_regularization = 0.1).fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7212    5]
 [ 157  163]]
              precision    recall  f1-score   support

         0.0       0.98      1.00      0.99      7217
         1.0       0.97      0.51      0.67       320

    accuracy                           0.98      7537
   macro avg       0.97      0.75      0.83      7537
weighted avg       0.98      0.98      0.98      7537



In [34]:
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
clf1 = LogisticRegression(n_jobs = -1, random_state=0, max_iter = 1, solver = 'saga')
clf2 = RandomForestClassifier(n_estimators=25, random_state=0, bootstrap = False)
clf3 = GaussianNB()
eclf1 = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)], voting='hard')
eclf1 = eclf1.fit(X_train, y_train)

eclf2 = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],voting='soft')
eclf2 = eclf2.fit(X_train, y_train)

eclf3 = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],voting='soft', weights=[2,1,1],flatten_transform=True)
eclf3 = eclf3.fit(X_train, y_train)

In [295]:
y_test_pred = eclf2.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7154   63]
 [ 176  144]]
              precision    recall  f1-score   support

         0.0       0.98      0.99      0.98      7217
         1.0       0.70      0.45      0.55       320

    accuracy                           0.97      7537
   macro avg       0.84      0.72      0.77      7537
weighted avg       0.96      0.97      0.97      7537



In [245]:
grid_predictions = eclf2.predict(test1)
pred=pd.DataFrame(grid_predictions, index = test.id)
pred.value_counts()

0.0    12029
1.0      345
dtype: int64

In [246]:
import os
os.chdir(r'../data')

pred.to_csv(r"submission_vot.csv")

In [301]:
from sklearn.multiclass import OutputCodeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
clf = OutputCodeClassifier(estimator=RandomForestClassifier(n_estimators=25, random_state=0, bootstrap = False)).fit(X_train, y_train)
y_test_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7198   19]
 [ 209  111]]
              precision    recall  f1-score   support

         0.0       0.97      1.00      0.98      7217
         1.0       0.85      0.35      0.49       320

    accuracy                           0.97      7537
   macro avg       0.91      0.67      0.74      7537
weighted avg       0.97      0.97      0.96      7537



### XGBoost

In [24]:
import xgboost as xgb
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [22]:
xg = xgb.XGBClassifier(verbosity=0, colsample_bytree = 0.9, use_label_encoder=False, n_estimators = 175, 
                      scale_pos_weight = 75).fit(X_train, y_train)

In [22]:
y_test_pred = xg.predict(X_test)
print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[7180   37]
 [ 111  209]]
              precision    recall  f1-score   support

         0.0       0.98      0.99      0.99      7217
         1.0       0.85      0.65      0.74       320

    accuracy                           0.98      7537
   macro avg       0.92      0.82      0.86      7537
weighted avg       0.98      0.98      0.98      7537



In [366]:
grid_predictions = xg.predict(test1)
pred=pd.DataFrame(grid_predictions, index = test.id)
pred.value_counts()

0    11993
1      381
dtype: int64

In [367]:
import os
os.chdir(r'../data')

pred.to_csv(r"submission_xgb.csv")